File: real-estate-dashboard/src/types/notification.ts
/**
 * Notification Types
 * Vollständige Type-Definitionen für das Notification System
 */

export enum NotificationType {
  INFO = 'info',
  SUCCESS = 'success',
  WARNING = 'warning',
  ERROR = 'error',
  REMINDER = 'reminder',
}

export enum NotificationCategory {
  SYSTEM = 'system',
  PROPERTY = 'property',
  CONTACT = 'contact',
  TASK = 'task',
  APPOINTMENT = 'appointment',
  DOCUMENT = 'document',
  FINANCIAL = 'financial',
  MESSAGE = 'message',
  TEAM = 'team',
  CIM = 'cim',
}

export enum NotificationPriority {
  LOW = 'low',
  NORMAL = 'normal',
  HIGH = 'high',
  URGENT = 'urgent',
}

export interface Notification {
  id: string;
  type: NotificationType;
  category: NotificationCategory;
  priority: NotificationPriority;
  title: string;
  message: string;
  read: boolean;
  read_at: string | null;
  archived: boolean;
  archived_at: string | null;
  action_url: string | null;
  action_label: string | null;
  related_entity_type: string | null;
  related_entity_id: string | null;
  related_entity_title: string | null;
  metadata: Record<string, any>;
  expires_at: string | null;
  created_at: string;
  updated_at: string;
  created_by_id: string | null;
}

export interface NotificationStats {
  total: number;
  unread: number;
  by_category: Record<string, number>;
  by_priority: Record<string, number>;
  by_type: Record<string, number>;
}

export interface NotificationListResponse {
  items: Notification[];
  total: number;
  page: number;
  size: number;
  pages: number;
  has_next: boolean;
  has_prev: boolean;
  stats?: NotificationStats;
}

export interface NotificationCreate {
  user_id: string;
  type?: NotificationType;
  category?: NotificationCategory;
  priority?: NotificationPriority;
  title: string;
  message: string;
  action_url?: string;
  action_label?: string;
  related_entity_type?: string;
  related_entity_id?: string;
  related_entity_title?: string;
  metadata?: Record<string, any>;
  expires_at?: string;
}

export interface NotificationUpdate {
  read?: boolean;
  archived?: boolean;
}

export interface NotificationMarkAsRead {
  notification_ids: string[];
}

export interface NotificationBulkAction {
  notification_ids: string[];
  action: 'mark_read' | 'mark_unread' | 'archive' | 'delete';
}

export interface NotificationPreference {
  id: string;
  category: NotificationCategory;
  enabled: boolean;
  email_enabled: boolean;
  push_enabled: boolean;
  in_app_enabled: boolean;
  min_priority: NotificationPriority;
  quiet_hours_enabled: boolean;
  quiet_hours_start: string | null;
  quiet_hours_end: string | null;
  created_at: string;
  updated_at: string;
}

export interface NotificationPreferenceUpdate {
  category: NotificationCategory;
  enabled?: boolean;
  email_enabled?: boolean;
  push_enabled?: boolean;
  in_app_enabled?: boolean;
  min_priority?: NotificationPriority;
  quiet_hours_enabled?: boolean;
  quiet_hours_start?: string;
  quiet_hours_end?: string;
}

export interface NotificationFilter {
  read?: boolean;
  archived?: boolean;
  category?: NotificationCategory;
  priority?: NotificationPriority;
  type?: NotificationType;
  from_date?: string;
  to_date?: string;
}

// UI Helpers
export interface NotificationIconConfig {
  icon: string;
  color: string;
  bgColor: string;
}

export const NOTIFICATION_TYPE_CONFIG: Record<NotificationType, NotificationIconConfig> = {
  [NotificationType.INFO]: {
    icon: 'ri-information-line',
    color: 'text-blue-600 dark:text-blue-400',
    bgColor: 'bg-blue-100 dark:bg-blue-900/30',
  },
  [NotificationType.SUCCESS]: {
    icon: 'ri-checkbox-circle-line',
    color: 'text-green-600 dark:text-green-400',
    bgColor: 'bg-green-100 dark:bg-green-900/30',
  },
  [NotificationType.WARNING]: {
    icon: 'ri-alert-line',
    color: 'text-orange-600 dark:text-orange-400',
    bgColor: 'bg-orange-100 dark:bg-orange-900/30',
  },
  [NotificationType.ERROR]: {
    icon: 'ri-error-warning-line',
    color: 'text-red-600 dark:text-red-400',
    bgColor: 'bg-red-100 dark:bg-red-900/30',
  },
  [NotificationType.REMINDER]: {
    icon: 'ri-alarm-line',
    color: 'text-purple-600 dark:text-purple-400',
    bgColor: 'bg-purple-100 dark:bg-purple-900/30',
  },
};

export const NOTIFICATION_CATEGORY_CONFIG: Record<NotificationCategory, { label: string; icon: string }> = {
  [NotificationCategory.SYSTEM]: { label: 'System', icon: 'ri-settings-3-line' },
  [NotificationCategory.PROPERTY]: { label: 'Immobilien', icon: 'ri-building-line' },
  [NotificationCategory.CONTACT]: { label: 'Kontakte', icon: 'ri-user-line' },
  [NotificationCategory.TASK]: { label: 'Aufgaben', icon: 'ri-task-line' },
  [NotificationCategory.APPOINTMENT]: { label: 'Termine', icon: 'ri-calendar-line' },
  [NotificationCategory.DOCUMENT]: { label: 'Dokumente', icon: 'ri-file-text-line' },
  [NotificationCategory.FINANCIAL]: { label: 'Finanzen', icon: 'ri-money-euro-circle-line' },
  [NotificationCategory.MESSAGE]: { label: 'Nachrichten', icon: 'ri-message-3-line' },
  [NotificationCategory.TEAM]: { label: 'Team', icon: 'ri-team-line' },
  [NotificationCategory.CIM]: { label: 'CIM', icon: 'ri-dashboard-line' },
};

export const NOTIFICATION_PRIORITY_CONFIG: Record<NotificationPriority, { label: string; color: string }> = {
  [NotificationPriority.LOW]: { label: 'Niedrig', color: 'text-gray-500' },
  [NotificationPriority.NORMAL]: { label: 'Normal', color: 'text-blue-500' },
  [NotificationPriority.HIGH]: { label: 'Hoch', color: 'text-orange-500' },
  [NotificationPriority.URGENT]: { label: 'Dringend', color: 'text-red-500' },
};

File: backend/app/schemas/notification.py
"""
Notification Pydantic Schemas
Request und Response Models für Notifications
"""
from typing import Optional, List, Dict, Any
from datetime import datetime, time
from pydantic import BaseModel, Field, ConfigDict
from enum import Enum


# Enums
class NotificationType(str, Enum):
    INFO = "info"
    SUCCESS = "success"
    WARNING = "warning"
    ERROR = "error"
    REMINDER = "reminder"


class NotificationCategory(str, Enum):
    SYSTEM = "system"
    PROPERTY = "property"
    CONTACT = "contact"
    TASK = "task"
    APPOINTMENT = "appointment"
    DOCUMENT = "document"
    FINANCIAL = "financial"
    MESSAGE = "message"
    TEAM = "team"
    CIM = "cim"


class NotificationPriority(str, Enum):
    LOW = "low"
    NORMAL = "normal"
    HIGH = "high"
    URGENT = "urgent"


# Request Models
class NotificationCreate(BaseModel):
    """Create notification request"""
    user_id: str
    type: NotificationType = NotificationType.INFO
    category: NotificationCategory = NotificationCategory.SYSTEM
    priority: NotificationPriority = NotificationPriority.NORMAL
    title: str = Field(..., max_length=255)
    message: str
    action_url: Optional[str] = Field(None, max_length=500)
    action_label: Optional[str] = Field(None, max_length=100)
    related_entity_type: Optional[str] = Field(None, max_length=50)
    related_entity_id: Optional[str] = Field(None, max_length=100)
    related_entity_title: Optional[str] = Field(None, max_length=255)
    metadata: Dict[str, Any] = Field(default_factory=dict)
    expires_at: Optional[datetime] = None


class NotificationUpdate(BaseModel):
    """Update notification request"""
    read: Optional[bool] = None
    archived: Optional[bool] = None


class NotificationBulkAction(BaseModel):
    """Bulk action request"""
    notification_ids: List[str]
    action: str = Field(..., pattern="^(mark_read|mark_unread|archive|delete)$")


class NotificationMarkAsRead(BaseModel):
    """Mark notification as read"""
    notification_ids: List[str]


# Response Models
class NotificationResponse(BaseModel):
    """Notification response model"""
    id: str
    type: NotificationType
    category: NotificationCategory
    priority: NotificationPriority
    title: str
    message: str
    read: bool
    read_at: Optional[datetime] = None
    archived: bool
    archived_at: Optional[datetime] = None
    action_url: Optional[str] = None
    action_label: Optional[str] = None
    related_entity_type: Optional[str] = None
    related_entity_id: Optional[str] = None
    related_entity_title: Optional[str] = None
    metadata: Dict[str, Any]
    expires_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    created_by_id: Optional[str] = None
    
    model_config = ConfigDict(from_attributes=True)


class NotificationStats(BaseModel):
    """Notification statistics"""
    total: int
    unread: int
    by_category: Dict[str, int]
    by_priority: Dict[str, int]
    by_type: Dict[str, int]


class NotificationListResponse(BaseModel):
    """Paginated notification list"""
    items: List[NotificationResponse]
    total: int
    page: int
    size: int
    pages: int
    has_next: bool
    has_prev: bool
    stats: Optional[NotificationStats] = None


# Notification Preference Models
class NotificationPreferenceUpdate(BaseModel):
    """Update notification preference"""
    category: NotificationCategory
    enabled: Optional[bool] = None
    email_enabled: Optional[bool] = None
    push_enabled: Optional[bool] = None
    in_app_enabled: Optional[bool] = None
    min_priority: Optional[NotificationPriority] = None
    quiet_hours_enabled: Optional[bool] = None
    quiet_hours_start: Optional[time] = None
    quiet_hours_end: Optional[time] = None


class NotificationPreferenceResponse(BaseModel):
    """Notification preference response"""
    id: str
    category: NotificationCategory
    enabled: bool
    email_enabled: bool
    push_enabled: bool
    in_app_enabled: bool
    min_priority: NotificationPriority
    quiet_hours_enabled: bool
    quiet_hours_start: Optional[time] = None
    quiet_hours_end: Optional[time] = None
    created_at: datetime
    updated_at: datetime
    
    model_config = ConfigDict(from_attributes=True)


class NotificationPreferencesBulk(BaseModel):
    """Bulk update notification preferences"""
    preferences: List[NotificationPreferenceUpdate]


# Filter Models
class NotificationFilter(BaseModel):
    """Notification filter options"""
    read: Optional[bool] = None
    archived: Optional[bool] = None
    category: Optional[NotificationCategory] = None
    priority: Optional[NotificationPriority] = None
    type: Optional[NotificationType] = None
    from_date: Optional[datetime] = None
    to_date: Optional[datetime] = None
